source/tomopy/misc/corr.py

```diff
@@ -267,6 +268,114 @@ def median_filter_cuda(arr, size=3, axis=0):
     return out
 
 
+def _determine_nonfinite_kernel_idxs(x_idx, y_idx, kernel, shape_x, shape_y):
+    """
+    Determine the proper kernel bounds for a given x, y index, and kernel size.
+
+    Parameters
+    ----------
+    x_idx : int
+        The x index for a given nonfinite value.
+    y_idx :  int
+        The y index for a given nonfinite value.
+    shape_x : int
+        The max shape of the 2D projection in the x dimension.
+    shape_y : int
+        The max shape of the 2D projection in the y dimension.
+
+    Returns
+    -------
+    x_lower, x_higher, y_lower, y_higher : int
+        The integer kernel bounds to surround the nonfinite value with.
+    """
+
+    # Determining x kernel
+    x_idx_lower = x_idx - kernel
+    x_idx_higher = x_idx + kernel + 1
+
+    if x_idx_lower < 0:
+        x_idx_lower = 0
+    if x_idx_higher > shape_x:
+        x_idx_higher = shape_x
+
+    # Determining y kernel
+    y_idx_lower = y_idx - kernel
+    y_idx_higher = y_idx + kernel + 1
+
+    if y_idx_lower < 0:
+        y_idx_lower = 0
+    if y_idx_higher > shape_y:
+        y_idx_higher = shape_y
+
+    return x_idx_lower, x_idx_higher, y_idx_lower, y_idx_higher
+
+
+def median_filter_nonfinite(data, size=3, callback=None):
+    """
+    Remove nonfinite values from a 3D array using an in-place 2D median filter.
+
+    The 2D selective median filter is applied along the last two axes of
+    the array.
+
+    Parameters
+    ----------
+    data : ndarray
+        The 3D array of data with nonfinite values in it.
+    size : int, optional
+        The size of the filter.
+    callback : func(total, description, unit)
+        A function called after every internal loop iteration.
+        total is number of loop iterations.
+        description is 'Nonfinite median filter'.
+        unit is ' prjs'.
+
+    Returns
+    -------
+    ndarray
+        The corrected 3D array with all nonfinite values removed based upon the local
+        median value defined by the kernel size.
+    """
+    # Defining a callback function if None is provided
+    if callback is None:
+
+        def callback(total, description, unit):
+            pass
+
+    # Iterating throug each projection to save on RAM
+    for projection in data:
+        nonfinite_idx = np.nonzero(~np.isfinite(projection))
+
+        # Iterating through each bad value and replace it with finite median
+        for x_idx, y_idx in zip(*nonfinite_idx):
+
+            # Determining the lower and upper bounds for kernel
+            (
+                x_lower,
+                x_higher,
+                y_lower,
+                y_higher,
+            ) = _determine_nonfinite_kernel_idxs(
+                x_idx,
+                y_idx,
+                size // 2,
+                data.shape[1],
+                data.shape[2],
+            )
```

Comment:
Once refactored to just 4 lines,  `_determine_nonfinite_kernel_idxs` can be removed and inlined here.